File: java/BillsApi/src/main/java/com/apilangarena/javabillsapi/application/bills/create/CreateBillService.java
package com.apilangarena.javabillsapi.application.bills.create;

import com.apilangarena.javabillsapi.application.common.ConflictException;
import com.apilangarena.javabillsapi.domain.bills.NewBill;
import com.apilangarena.javabillsapi.domain.bills.NewBillLine;
import org.springframework.dao.DataIntegrityViolationException;
import org.springframework.stereotype.Service;
import org.springframework.transaction.support.TransactionTemplate;

import java.time.OffsetDateTime;
import java.time.ZoneOffset;
import java.util.ArrayList;
import java.util.List;

@Service
public class CreateBillService {
    private final TransactionTemplate transactionTemplate;
    private final BillWriteRepository billWriteRepository;
    private final IntegrationEventPublisher integrationEventPublisher;

    public CreateBillService(
        TransactionTemplate transactionTemplate,
        BillWriteRepository billWriteRepository,
        IntegrationEventPublisher integrationEventPublisher
    ) {
        this.transactionTemplate = transactionTemplate;
        this.billWriteRepository = billWriteRepository;
        this.integrationEventPublisher = integrationEventPublisher;
    }

    public CreateBillResult execute(CreateBillCommand command) {
        CreateBillResult created = transactionTemplate.execute(status -> {
            String normalizedBillNumber = command.billNumber() == null ? "" : command.billNumber().trim();
            if (billWriteRepository.existsByBillNumber(normalizedBillNumber)) {
                throw new ConflictException("Bill number '" + command.billNumber() + "' already exists.");
            }

            List<NewBillLine> lines = new ArrayList<>();
            for (int idx = 0; idx < command.lines().size(); idx++) {
                CreateBillLineCommand line = command.lines().get(idx);
                lines.add(NewBillLine.create(
                    idx + 1,
                    line.concept(),
                    line.quantity(),
                    line.unitAmount()
                ));
            }

            NewBill newBill = NewBill.create(
                command.billNumber(),
                command.issuedAt(),
                command.customerName(),
                command.currency(),
                command.tax(),
                lines
            );

            try {
                return billWriteRepository.create(newBill);
            } catch (DataIntegrityViolationException ignored) {
                throw new ConflictException("Bill number '" + command.billNumber() + "' already exists.");
            }
        });

        if (created == null) {
            throw new IllegalStateException("Transaction did not produce a created bill result.");
        }

        integrationEventPublisher.publishBillCreated(
            new BillCreatedEvent(
                created.id(),
                created.billNumber(),
                created.issuedAt(),
                created.subtotal(),
                created.tax(),
                created.total(),
                created.currency(),
                OffsetDateTime.now(ZoneOffset.UTC),
                "java-api"
            )
        );

        return created;
    }
}

File: kotlin/BillsApi/src/main/kotlin/com/apilangarena/billsapi/application/bills/create/CreateBillService.kt
package com.apilangarena.billsapi.application.bills.create

import com.apilangarena.billsapi.application.common.ConflictException
import com.apilangarena.billsapi.domain.bills.NewBill
import com.apilangarena.billsapi.domain.bills.NewBillLine
import org.springframework.dao.DataIntegrityViolationException
import org.springframework.stereotype.Service
import org.springframework.transaction.support.TransactionTemplate
import java.time.OffsetDateTime

@Service
class CreateBillService(
    private val transactionTemplate: TransactionTemplate,
    private val billWriteRepository: BillWriteRepository,
    private val integrationEventPublisher: IntegrationEventPublisher,
) {
    fun execute(command: CreateBillCommand): CreateBillResult {
        val created = transactionTemplate.execute {
            if (billWriteRepository.existsByBillNumber(command.billNumber.trim())) {
                throw ConflictException("Bill number '${command.billNumber}' already exists.")
            }

            val lines = command.lines.mapIndexed { idx, line ->
                NewBillLine.create(
                    lineNo = idx + 1,
                    concept = line.concept,
                    quantity = line.quantity,
                    unitAmount = line.unitAmount,
                )
            }

            val newBill = NewBill.create(
                billNumber = command.billNumber,
                issuedAt = command.issuedAt,
                customerName = command.customerName,
                currency = command.currency,
                tax = command.tax,
                lines = lines,
            )

            try {
                billWriteRepository.create(newBill)
            } catch (_: DataIntegrityViolationException) {
                throw ConflictException("Bill number '${command.billNumber}' already exists.")
            }
        } ?: throw IllegalStateException("Transaction did not produce a created bill result.")

        // Publish only after transaction commit has completed.
        integrationEventPublisher.publishBillCreated(
            BillCreatedEvent(
                billId = created.id,
                billNumber = created.billNumber,
                issuedAt = created.issuedAt,
                subtotal = created.subtotal,
                tax = created.tax,
                total = created.total,
                currency = created.currency,
                occurredAtUtc = OffsetDateTime.now(),
                source = "kotlin-api",
            )
        )

        return created
    }
}
